src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs

```diff
@@ -103,36 +133,42 @@ internal static void Traverse<T>(this SelectExpandClause selectExpandClause, Fun
             {
-                string currentExpandClause = String.Join("/", expandSelectItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray());
-                T subResult = default(T);
-                if (expandSelectItem.SelectAndExpand.SelectedItems.Any())
-                {
-                    Traverse(expandSelectItem.SelectAndExpand, processSubResult, combineSelectAndExpand, processApply, out subResult);
-                }
-
-                if (expandSelectItem.ApplyOption != null && processApply != null)
-                {
-                    subResult = processApply(expandSelectItem.ApplyOption);
-                }
+                // $expand=..../$ref
+                ExpandedReferenceSelectItem expandRefItem = selectItem as ExpandedReferenceSelectItem;
+                ExpandedNavigationSelectItem expandSelectItem = selectItem as ExpandedNavigationSelectItem;
 
-                var expandItem = processSubResult(currentExpandClause, subResult);
-                if (expandItem != null)
+                if (expandSelectItem != null)
                 {
-                    expandList.Add(expandItem);
-                }
-            }
+                    string currentExpandClause = String.Join("/", expandSelectItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray());
+                    T subResult = default(T);
+                    if (expandSelectItem.SelectAndExpand.SelectedItems.Any())
+                    {
+                        Traverse(expandSelectItem.SelectAndExpand, processSubResult, combineSelectAndExpand, processApply, out subResult);
+                    }
 
-            foreach (ExpandedReferenceSelectItem expandSelectItem in selectExpandClause.SelectedItems.Where(I => I.GetType() == typeof(ExpandedReferenceSelectItem)))
-            {
-                string currentExpandClause = String.Join("/", expandSelectItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray());
-                currentExpandClause += "/$ref";
+                    if (expandSelectItem.ApplyOption != null && processApply != null)
+                    {
+                        subResult = processApply(expandSelectItem.ApplyOption);
+                    }
 
-                var expandItem = processSubResult(currentExpandClause, default(T));
-                if (expandItem != null)
+                    var expandItem = processSubResult(currentExpandClause, subResult);
+                    if (expandItem != null)
+                    {
+                        expandList.Add(expandItem);
+                    }
+                }
+                else if (expandRefItem != null)
                 {
-                    expandList.Add(expandItem);
+                    string currentExpandClause = String.Join("/", expandRefItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray());

```

Comment:
>expandRefItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray() [](start = 66, length = 97)

Converting ToArray and then joining is expensive and requires allocations. Can you create your own private method to loop through and create a string through string builder.?

Reply:
Current .NET version string join takes Ienumerable whereas older versions needs array. Other way is to use StringBuilder, Append and return .ToString(). But as of now not sure whether that would give any perf improvement. We can figure out and if it does can have a custom string join method for all string joins